vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,230 @@
+use futures::*;
+use snow::Session::{self, Handshake};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+const NOISE_MAX_SIZE: usize = 65535;
+
+#[derive(DebugStub)]
+pub enum Step {
+    Read {
+        #[debug_stub = "encrypted buffer"]
+        enc_buffer: [u8; NOISE_MAX_SIZE],
+        len: usize,
+    },
+    Write {
+        #[debug_stub = "buffer"]
+        buffer: [u8; NOISE_MAX_SIZE],
+        remaining_bytes: usize,
+        len: usize,
+    },
+}
+
+impl Step {
+    fn read() -> Self {
+        Step::Read {
+            enc_buffer: [0u8; NOISE_MAX_SIZE],
+            len: 0,
+        }
+    }
+
+    fn write() -> Self {
+        Step::Write {
+            buffer: [0u8; NOISE_MAX_SIZE],
+            remaining_bytes: 0,
+            len: 0,
+        }
+    }
+}
+
+pub trait InitHandshake {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite;
+}
+
+impl InitHandshake for Session {
+    fn init_hand_shake<IO>(self, io: IO) -> NoiseHandshake<IO>
+    where
+        IO: AsyncRead + AsyncWrite,
+    {
+        match self {
+            Handshake(ref handshake_state) => NoiseHandshake {
+                next: if handshake_state.is_initiator() {
+                    Step::write()
+                } else {
+                    Step::read()
+                },
+                noise: Some(self),
+                io: Some(io),
+            },
+            _ => panic!(
+                "Noise Session in incorrect state, you should init before starting handshake"
+            ),
+        }
+    }
+}
+
+#[derive(Debug)]
```

Comment:
:wrench: Futures don't usually have to be `Debug` because you are very likely going to spawn them anyway and then you no longer have a reference to them that you could print.